llvm/lib/Target/EraVM/EraVMInstrInfo.cpp

```diff
@@ -417,6 +417,71 @@ unsigned EraVMInstrInfo::insertBranch(
   return Count;
 }
 
+bool EraVMInstrInfo::analyzeCompare(const MachineInstr &MI, Register &SrcReg,
+                                    Register &SrcReg2, int64_t &CmpMask,
+                                    int64_t &CmpValue) const {
+  return isSub(MI) && isFlagSettingInstruction(MI);
+}
+
+bool EraVMInstrInfo::optimizeCompareInstr(
+    MachineInstr &CmpInstr, Register SrcReg, Register SrcReg2, int64_t CmpMask,
+    int64_t CmpValue, const MachineRegisterInfo *MRI) const {
+  assert(CmpInstr.getParent() && "CmpInstr must be in a basic block");
+  assert(MRI && "MachineRegisterInfo is required");
+
+  // If Flags register is not used, convert this compare instruction (sub
+  // instruction that sets flags) to a sub that doesn't set flags.
+  int DeadFlagsIdx = CmpInstr.findRegisterDefOperandIdx(EraVM::Flags, true);
+  if (DeadFlagsIdx != -1) {
+    int NewOpc = EraVM::getNonFlagSettingOpcode(CmpInstr.getOpcode());
+    assert(NewOpc != -1 && "Invalid opcode for compare instruction");
+    CmpInstr.setDesc(get(NewOpc));
+    CmpInstr.removeOperand(DeadFlagsIdx);
+    return true;
+  }
+
+  // TODO: #621 Support removing compare instruction which output is used if the
+  // nearest flag setting instruction is identical.
+  if (!MRI->use_nodbg_empty(CmpInstr.getOperand(0).getReg()))
+    return false;
+
+  // In some cases, we can have the following sequence:
+  //   Out1 = SUB_v In1, In2, CondCC1, implicit-def $flags
+  //   ...
+  //   Out2 = SUB_v In3, In4, CondCC2, implicit-def dead $flags
+  //   ...
+  //   Out3 = SUB_v In1, In2, CondCC1, implicit-def $flags
+  //
+  // In this case, `Out2 = SUB_v` that will be converted to _s in this function,
+  // prevents MachineCSE to remove `Out3 = SUB_v` in favor of `Out1 = SUB_v`,
+  // just because of `implicit-def dead $flags`, so try to find identical
+  // compare instruction and remove it here.
+  //
+  // Search backwards in order to find the nearest flag setting instruction that
+  // can be identical to this compare instruction.
+  auto From = std::next(MachineBasicBlock::reverse_iterator(CmpInstr));
+  for (MachineInstr &MI : make_range(From, CmpInstr.getParent()->rend())) {
+    if (isFlagSettingInstruction(MI)) {
```

Comment:
Perhaps we should redefine `isFlagSettingInstruction` as
```
any_of(MI.implicit_operands(), [](const MachineOperand &MO) {
          return MO.isReg() && MO.isDef() && MO.getReg() == EraVM::Flags;
        }))
```
And then use `find_if` instead of loop.

Also from the logic it looks like that by the time we do this search, `Out2 = SUB_v ...` is already converted and the `impact-def` is dropped while the text about says that the instruction will be converted. So could you clarify on that?

Reply:
Nice catch. Added `find_if` to simplify the code and added clarification that we can now do the CSE since we dropped `impact-def` for `Out2 = SUB_v ...`.